readium/streamer/src/main/java/org/readium/r2/streamer/PublicationFactory.kt

```diff
@@ -0,0 +1,267 @@
+     *   publication, for example a password.
+     * @param allowUserInteraction Indicates whether the user can be prompted, for example for its
+     *   credentials.
+     * @param sender Free object that can be used by reading apps to give some UX context when
+     *   presenting dialogs.
+     * @param onCreatePublication Transformation which will be applied on the Publication Builder.
+     *   It can be used to modify the [Manifest], the root [Fetcher] or the list of service
+     *   factories of the [Publication].
+     * @param warnings Logger used to broadcast non-fatal parsing warnings.
+     * @return A [Publication] or a [Publication.OpeningException] in case of failure.
+     */
+    suspend fun open(
+        asset: Asset,
+        drmScheme: String? = null,
+        credentials: String? = null,
+        allowUserInteraction: Boolean,
+        sender: Any? = null,
+        onCreatePublication: Publication.Builder.() -> Unit = {},
+        warnings: WarningLogger? = null
+    ): PublicationTry<Publication> {
+        val compositeOnCreatePublication: Publication.Builder.() -> Unit = {
+            this@PublicationFactory.onCreatePublication(this)
+            onCreatePublication(this)
+        }
+
+        return if (drmScheme == null) {
+            openUnprotected(
+                asset,
+                compositeOnCreatePublication,
+                warnings
+            )
+        } else {
+            openProtected(
+                asset,
+                drmScheme,
+                credentials,
+                allowUserInteraction,
+                sender,
+                compositeOnCreatePublication,
+                warnings
+            )
+        }
+    }
+
+    private suspend fun openUnprotected(
+        asset: Asset,
+        onCreatePublication: Publication.Builder.() -> Unit,
+        warnings: WarningLogger?
+    ): Try<Publication, Publication.OpeningException> {
+        val parserAsset = parserAssetFactory.createParserAsset(asset)
+            .getOrElse { return Try.failure(it) }
+        return openParserAsset(parserAsset, onCreatePublication, warnings)
+    }
+
+    private suspend fun openProtected(
+        asset: Asset,
+        drmScheme: String,
+        credentials: String?,
+        allowUserInteraction: Boolean,
+        sender: Any?,
+        onCreatePublication: Publication.Builder.() -> Unit,
+        warnings: WarningLogger?
+    ): Try<Publication, Publication.OpeningException> {
+        val protectedAsset = contentProtections
+            .lazyMapFirstNotNullOrNull {
+                it.open(asset, drmScheme, credentials, allowUserInteraction, sender)
+            }?.getOrElse { return Try.failure(it) }
+            ?: return Try.failure(Publication.OpeningException.Forbidden())
+
+        val parserAsset = PublicationParser.Asset(
+            protectedAsset.name,
+            protectedAsset.mediaType,
+            protectedAsset.fetcher
+        )
+
```

Comment:
Seeing this comforts me in the idea of having a shared `PublicationAsset` type. That would simplify the chaining of `unlock(Asset) -> PublicationAsset` and then `open(PublicationAsset) -> Publication`.